File: examples/native-counter/background.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>

int main(void) {
    const char *data = getenv("DEVOS_DATA_DIR");
    if (!data || chdir(data) != 0) return 1;
    unsigned long ticks = 0;
    FILE *file = fopen("counter.json", "r");
    if (file) {
        if (fscanf(file, "{\"ticks\": %lu", &ticks) != 1) ticks = 0;
        fclose(file);
    }
    for (;;) {
        file = fopen("counter.new", "w");
        if (!file) return 1;
        if (fprintf(file, "{\"ticks\": %lu, \"updated\": %lld}\n", ++ticks,
                    (long long)time(NULL)) < 0) { fclose(file); return 1; }
        if (fclose(file) != 0 || rename("counter.new", "counter.json") != 0) return 1;
        sleep(1);
    }
}

File: examples/desktop-counter/payload/opt/apps/desktop-counter/background.py
#!/usr/bin/python3
import json
import os
from pathlib import Path
import time

data = Path(os.environ['DEVOS_DATA_DIR'])
path = data / 'counter.json'
try:
    ticks = int(json.loads(path.read_text())['ticks'])
except (OSError, ValueError, KeyError):
    ticks = 0
while True:
    ticks += 1
    temporary = data / 'counter.new'
    temporary.write_text(json.dumps({'ticks': ticks, 'updated': time.time()}))
    temporary.replace(path)
    time.sleep(1)
